### ISSUE TITLE
issue(sqlite3): WAL - journal_mode and multi-thread support - err(522) SQLITE_IOERR_SHORT_READ

### ISSUE THREAD

**Issue by rinor** [nanovms/nanos#1911]
Is there any known issue/gotcha to be aware of while testing? (i.e: not supported yet fs/kernel stuff)

**Comment by eyberg** [comment]
i'm not aware of anything atm, if this needs more discussion perhaps we can mv it to the discussions area? - if you do come across anything feel free to open an issue

**Comment by rinor** [comment]
I'm about to start testing in a few, and if issues found, the communication thread may be similar to https://github.com/nanovms/nanos/issues/1845.

If you prefer to have separate gh issue for each problem found, feel free to close this.

**Comment by rinor** [comment]
<details><summary><code>Code</code> used to test</summary>

  - binary named `nanos-sqlite`

```sh
$ go mod init nanos-sqlite
$ go mod tidy
$ CGO_ENABLED=0 GO111MODULE=on go build --ldflags "-s -w" -trimpath -a -v -tags osusergo,netgo
```



```go
package main

import (
	"database/sql"
	"log"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"

	_ "modernc.org/sqlite"
	//
	// _ "github.com/mattn/go-sqlite3" // to be tested next (requires cgo)
)

const (
	NANOS_SQLITE3_WRITE_CONNECTION_ENV     = "NANOS_SQLITE3_WRITE_CONNECTION"
	NANOS_SQLITE3_WRITE_CONNECTION_DEFAULT = "file:" + "nanos.sqlite3" + "?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000"

	NANOS_SQLITE3_READ_CONNECTION_ENV = "NANOS_SQLITE3_READ_CONNECTION"

	NANOS_SQLITE3_MAX_GOROUTINES_WRITE_ENV     = "NANOS_SQLITE3_MAX_GOROUTINES_WRITE"
	NANOS_SQLITE3_MAX_GOROUTINES_WRITE_DEFAULT = 1

	NANOS_SQLITE3_MAX_GOROUTINES_READ_ENV     = "NANOS_SQLITE3_MAX_GOROUTINES_READ"
	NANOS_SQLITE3_MAX_GOROUTINES_READ_DEFAULT = 1

	NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE_ENV     = "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE"
	NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE_DEFAULT = 100

	NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ_ENV     = "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ"
	NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ_DEFAULT = 100

	NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS_ENV     = "NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS"
	NANOS_SQLITE3_RW_RANDON_WAIT_MILLIS_DEFAULT = 0

	NANOS_SQLITE3_READ_ALL_DATA_ENV = "NANOS_SQLITE3_READ_ALL_DATA"

	NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS_ENV = "NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS"
)

const (
	sql_create_table   = "CREATE TABLE IF NOT EXISTS `nanos` (`data` VARCHAR(64) NULL);"
	sql_insert_prepare = "INSERT INTO nanos(data) values(?)"
	sql_select_prepare = "SELECT rowid, data FROM nanos WHERE rowid = ? LIMIT 1"
)

func main() {
	/* just wait for nanos to complete and output ipv6 on local qemu */
	time.Sleep(time.Duration(2) * time.Second)

	defer func() {
		log.Println()
		log.Printf("DEFER - MAIN - ...will terminate in %ds\n", 5)
		time.Sleep(time.Duration(5) * time.Second)
		log.Println("DEFER - MAIN - DONE...")
	}()

	var (
		sql_write_connection string = NANOS_SQLITE3_WRITE_CONNECTION_DEFAULT
		sql_read_connection  string = sql_write_connection

		max_conns_write int = NANOS_SQLITE3_MAX_GOROUTINES_WRITE_DEFAULT
		max_conns_read  int = NANOS_SQLITE3_MAX_GOROUTINES_READ_DEFAULT

		items_write_per_goroutine int = NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE_DEFAULT
		items_read_per_goroutine  int = NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ_DEFAULT

		rw_random_wait int = NANOS_SQLITE3_RW_RANDON_WAIT_MILLIS_DEFAULT

		sqlite_version   string
		sqlite_source_id string

		sqlite_use_multiple_connections bool

		wg sync.WaitGroup

		conns_pool_write chan *sql.Stmt
		conns_pool_read  chan *sql.Stmt
	)

	/*
		get envs or default
	*/
	{
		// use separate connection for each writer/reader
		sqlite_use_multiple_connections = os.Getenv(NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS_ENV) != ""

		// write connection string
		sql_write_connection_env := os.Getenv(NANOS_SQLITE3_WRITE_CONNECTION_ENV)
		if sql_write_connection_env != "" {
			sql_write_connection = sql_write_connection_env
			// set also read conn here
			sql_read_connection = sql_write_connection
		}

		// read connection string, only if sqlite_use_multiple_connections
		if sqlite_use_multiple_connections {
			sql_read_connection_env := os.Getenv(NANOS_SQLITE3_READ_CONNECTION_ENV)
			if sql_read_connection_env != "" {
				sql_read_connection = sql_read_connection_env
			}
		}

		// write max con pool
		max_conns_write_env := os.Getenv(NANOS_SQLITE3_MAX_GOROUTINES_WRITE_ENV)
		if max_conns_write_env != "" {
			max_val, err := strconv.Atoi(max_conns_write_env)
			if err != nil {
				log.Printf("strconv.Atoi(%s): %q\n", max_conns_write_env, err)
			} else if max_val >= 0 {
				max_conns_write = max_val
			}
		}

		// read max con pool
		max_conns_read_env := os.Getenv(NANOS_SQLITE3_MAX_GOROUTINES_READ_ENV)
		if max_conns_read_env != "" {
			max_val, err := strconv.Atoi(max_conns_read_env)
			if err != nil {
				log.Printf("strconv.Atoi(%s): %q\n", max_conns_read_env, err)
			} else if max_val >= 0 {
				max_conns_read = max_val
			}
		}

		// write max items per goroutine
		items_write_per_goroutine_env := os.Getenv(NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE_ENV)
		if items_write_per_goroutine_env != "" {
			items_val, err := strconv.Atoi(items_write_per_goroutine_env)
			if err != nil {
				log.Printf("strconv.Atoi(%s): %q\n", items_write_per_goroutine_env, err)
			} else if items_val >= 0 {
				items_write_per_goroutine = items_val
			}
		}

		// read max items per goroutine
		items_read_per_goroutine_env := os.Getenv(NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ_ENV)
		if items_read_per_goroutine_env != "" {
			items_val, err := strconv.Atoi(items_read_per_goroutine_env)
			if err != nil {
				log.Printf("strconv.Atoi(%s): %q\n", items_read_per_goroutine_env, err)
			} else if items_val >= 0 {
				items_read_per_goroutine = items_val
			}
		}

		// random max wait in millis on each r/w
		rw_random_wait_env := os.Getenv(NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS_ENV)
		if rw_random_wait_env != "" {
			val, err := strconv.Atoi(rw_random_wait_env)
			if err != nil {
				log.Printf("strconv.Atoi(%s): %q\n", rw_random_wait_env, err)
			} else if val > 0 {
				rw_random_wait = val
			}
		}
	}

	/*
		info
	*/
	{
		log.Println()
		log.Printf("Connection Write String: %q\n", sql_write_connection)
		log.Printf("Connection Read String : %q\n", sql_read_connection)
		log.Println()
		log.Printf("Use mutiple connections: %t\n", sqlite_use_multiple_connections)
		log.Println()
		log.Printf("Write goroutines max: %d\n", max_conns_write)
		log.Printf("Read  goroutines max: %d\n", max_conns_read)
		log.Println()
		log.Printf("Write items per goroutine: %d\n", items_write_per_goroutine)
		log.Printf("Read  items per goroutine: %d\n", items_read_per_goroutine)
		log.Println()
		log.Printf("R/W Random Wait Millis : [0,%d)\n", rw_random_wait)
		log.Printf("ReadAllData at the end : %t\n", os.Getenv(NANOS_SQLITE3_READ_ALL_DATA_ENV) != "")
		log.Println()
	}

	/*
		setup main db connection and tables
	*/
	db_main, err := sql.Open("sqlite", sql_write_connection)
	if err != nil {
		log.Panicf("SETUP: sql.Open(%s) err(%q)\n", sql_write_connection, err)
	}
	// deferred because
	// 1. if db is in :memory: it will delete everything for all connections once closed
	// 2. we may want to use single db connection
	defer func() {
		err_db := db_main.Close()
		log.Printf("DEFER - SETUP - db(%p) err(%q) - closed\n", db_main, err_db)
	}()

	// make sure we can connect
	err = db_main.Ping()
	if err != nil {
		log.Panicf("SETUP: db_main.Ping(%s) err(%q)\n", sql_write_connection, err)
	}

	// create table
	_, err = db_main.Exec(sql_create_table)
	if err != nil {
		log.Panicf("SETUP: db.Exec(%s)  err(%q)\n", sql_create_table, err)
	}
	/*
		setup main db connection stmt(s)
	*/
	// write stmt
	stmt_main_insert, err := db_main.Prepare(sql_insert_prepare)
	if err != nil {
		log.Panicf("SETUP: db.Prepare(%s) err(%q)\n", sql_insert_prepare, err)
	}
	defer func() {
		err_stmt := stmt_main_insert.Close()
		log.Printf("DEFER - SETUP - stmt_main_insert(%p) err(%q) - closed\n", stmt_main_insert, err_stmt)
	}()
	// read stmt
	stmt_main_select, err := db_main.Prepare(sql_select_prepare)
	if err != nil {
		log.Panicf("SETUP: db.Prepare(%s) err(%q)\n", sql_select_prepare, err)
	}
	defer func() {
		err_stmt := stmt_main_select.Close()
		log.Printf("DEFER - SETUP - stmt_main_select(%p) err(%q)- closed\n", stmt_main_select, err_stmt)
	}()

	/*
		info sqlite3 version - from main connection
	*/
	// sqlite version
	_ = db_main.QueryRow("SELECT SQLITE_VERSION()").Scan(&sqlite_version)
	log.Printf("SQLITE_VERSION()  : %s\n", sqlite_version)
	// sqlite source version
	_ = db_main.QueryRow("SELECT SQLITE_SOURCE_ID()").Scan(&sqlite_source_id)
	log.Printf("SQLITE_SOURCE_ID(): %s\n", sqlite_source_id)
	log.Println()

	/*
		setup write pool - 1 db con with 1 prepared stmt per write worker - if sqlite_use_multiple_connections
	*/
	if sqlite_use_multiple_connections {
		conns_pool_write = make(chan *sql.Stmt, max_conns_write)
		for i := 0; i < max_conns_write; i++ {

			dbw, err := sql.Open("sqlite", sql_write_connection)
			if err != nil {
				log.Panicf("WRITE: sql.Open(%s)-(%d) err(%q)\n", sql_write_connection, i, err)

			}
			defer func(i int) {
				err_db := dbw.Close()
				log.Printf("DEFER - WRITE - db(%d - %p) err(%q) - closed\n", i, dbw, err_db)
			}(i)

			err = dbw.Ping()
			if err != nil {
				log.Panicf("WRITE: dbw.Ping(%s)-(%d) err(%q)\n", sql_write_connection, i, err)
			}

			stmt, err := dbw.Prepare(sql_insert_prepare)
			if err != nil {
				log.Panicf("WRITE: db.Prepare(%s)-(%d): err(%q)\n", sql_insert_prepare, i, err)
			}
			defer func(i int) {
				err_stmt := stmt.Close()
				log.Printf("DEFER - WRITE - stmt(%d - %p) err(%q) - closed\n", i, stmt, err_stmt)
			}(i)

			conns_pool_write <- stmt
		}
	}
	// get a stmt from the pool if sqlite_use_multiple_connections, otherwise return stmt_main_insert
	checkout_write_conn := func() *sql.Stmt {
		if sqlite_use_multiple_connections {
			return <-conns_pool_write
		}
		return stmt_main_insert
	}
	// return a stmt to the pool if sqlite_use_multiple_connections, otherwise nothing to do
	checkin_write_conn := func(c *sql.Stmt) {
		if sqlite_use_multiple_connections {
			conns_pool_write <- c
		}
	}

	/*
		setup read pool - 1 db con with 1 prepared stmt per read worker - if sqlite_use_multiple_connections
	*/
	if sqlite_use_multiple_connections {
		conns_pool_read = make(chan *sql.Stmt, max_conns_read)
		for i := 0; i < max_conns_read; i++ {

			dbr, err := sql.Open("sqlite", sql_read_connection)
			if err != nil {
				log.Panicf("READ: sql.Open(%s)-(%d) err(%q)\n", sql_read_connection, i, err)
			}
			defer func(i int) {
				err_db := dbr.Close()
				log.Printf("DEFER - READ - db(%d - %p) err(%q) - closed\n", i, dbr, err_db)
			}(i)
			err = dbr.Ping()
			if err != nil {
				log.Panicf("WRITE: dbr.Ping(%s)-(%d) err(%q)\n", sql_read_connection, i, err)
			}

			stmt, err := dbr.Prepare(sql_select_prepare)
			if err != nil {
				log.Panicf("READ: db.Prepare(%s)-(%d) err(%q)\n", sql_select_prepare, i, err)
			}
			defer func(i int) {
				err_stmt := stmt.Close()
				log.Printf("DEFER - READ - stmt(%d - %p) err(%q) - closed\n", i, stmt, err_stmt)
			}(i)

			conns_pool_read <- stmt
		}
	}
	// get a stmt from the pool if sqlite_use_multiple_connections,  otherwise return stmt_main_select
	checkout_read_conn := func() *sql.Stmt {
		if sqlite_use_multiple_connections {
			return <-conns_pool_read
		}
		return stmt_main_select
	}
	// return a stmt to the pool if sqlite_use_multiple_connections, otherwise nothing to do
	checkin_read_conn := func(c *sql.Stmt) {
		if sqlite_use_multiple_connections {
			conns_pool_read <- c
		}
	}

	log.Println("R/W goroutines work - STARTING")

	/*
		start write worker(s) - (connection is RW)
	*/
	for gw := 0; gw < max_conns_write; gw++ {
		wg.Add(1)

		go func(data string) {
			defer wg.Done()

			// pick one from the pool and return it when done
			stmt := checkout_write_conn()
			defer checkin_write_conn(stmt)

			for i := 0; i < items_write_per_goroutine; i++ {
				write_data := data + "-" + strconv.Itoa(i+1)
				_, err := stmt.Exec(write_data)
				if err != nil {
					log.Printf("WRITE: GW(%s) - stmt.Exec(%s) err(%q)\n", data, write_data, err)
					continue
				}
				if rw_random_wait > 0 {
					time.Sleep(time.Duration(rand.Intn(rw_random_wait)) * time.Millisecond)
				}
			}
		}("goroutine_" + strconv.Itoa(gw))
	}

	/*
		start read worker(s) - (connection is RW, unles RO configured at NANOS_SQLITE3_READ_CONNECTION and sqlite_use_multiple_connections)
	*/
	for gr := 0; gr < max_conns_read; gr++ {
		wg.Add(1)

		go func(gr int) {
			defer wg.Done()

			var (
				rowid int
				data  []byte
			)

			// pick one from the pool and return it when done
			stmt := checkout_read_conn()
			defer checkin_read_conn(stmt)

			for i := items_read_per_goroutine; i > 0; i-- {
				err := stmt.QueryRow(i).Scan(&rowid, &data)
				if err != nil {
					if err != sql.ErrNoRows {
						log.Printf("READ: GR(%d) - row.Scan(%d) err(%q)\n", gr, i, err)
					}
					continue
				}
				if rw_random_wait > 0 {
					time.Sleep(time.Duration(rand.Intn(rw_random_wait)) * time.Millisecond)
				}
			}
		}(gr)
	}

	// wait for worker(s) to be done
	wg.Wait()
	log.Println("R/W goroutines work - COMPLETED")

	/*
		get all data - if NANOS_SQLITE3_READ_ALL_DATA != ""
	*/
	if os.Getenv(NANOS_SQLITE3_READ_ALL_DATA_ENV) != "" {
		log.Println()
		log.Printf("Reading all data from nanos table...using separate connection - %t\n", sqlite_use_multiple_connections)

		db_read_all := db_main
		if sqlite_use_multiple_connections {
			db_read_all, err = sql.Open("sqlite", sql_read_connection)
			if err != nil {
				log.Panicf("ReadAllData: sql.Open(%s) err(%q)\n", sql_read_connection, err)
			}
			defer func() {
				err_db := db_read_all.Close()
				log.Printf("DEFER - ReadAllData - db(%p) err(%q) - closed\n", db_read_all, err_db)
			}()
		}

		rows, err := db_read_all.Query("SELECT rowid,data FROM nanos")
		if err != nil {
			log.Panicf("ReadAllData: db.Query(%s) err(%q)\n", "SELECT * FROM nanos", err)
		}
		defer func() {
			err_rows := rows.Close()
			log.Printf("DEFER - ReadAllData - rows err(%q) - closed\n", err_rows)
		}()

		var (
			rowid int
			data  []byte
		)
		for rows.Next() {
			err = rows.Scan(&rowid, &data)
			if err != nil {
				log.Panicf("ReadAllData: rows.Scan(%s) err(%q)\n", "SELECT * FROM nanos", err)
			}
			log.Printf("ReadAllData: %d - %s\n", rowid, data)
		}
	}
	log.Println()
}
```
</details>
  
---

While testing **WAL** mode on nanos,  was getting a lot of I/O errors (error code 522) under some setups that work perfectly fine on linux.

- (**522**) [SQLITE_IOERR_SHORT_READ](https://www.sqlite.org/rescode.html#ioerr_short_read)
  - The **SQLITE_IOERR_SHORT_READ** error code is an [extended error code](https://www.sqlite.org/rescode.html#pve) for [SQLITE_IOERR](https://www.sqlite.org/rescode.html#ioerr) indicating that a read attempt in the [VFS](https://www.sqlite.org/vfs.html) layer was unable to obtain as many bytes as was requested. This might be due to a truncated file.

Since the error mentions  [VFS](https://www.sqlite.org/vfs.html) and was using the default one `vfs=unix`, before investigating further, thought of testing/checking different unix `vfs` mechanisms just to compare the behavior with linux (not that I needed them).

---
And so picked the first one `vfs=unix-dotfile` from `unix-*` list:
- `unix-dotfile` - uses dot-file locking rather than POSIX advisory locks.

Unfortunately on **nanos** this causes a kernel assertion to fail on mutex ops, while operating on db data. 

<details><summary><b>ops</b> config</summary>

```sh
$ ops volume create nanos-sqlite -s 10G
$ ops run -c config_mutex.json nanos-sqlite --smp=1 -f -n
```

- `config_mutex.json`

```json
{
  "BaseVolumeSz": "32m",
  "Args": [],
  "Mounts": {
    "nanos-sqlite": "db"
  },
  "Dirs": [],
  "RunConfig": {
    "Memory": "8G"
  },
  "Env": {
    "NANOS_SQLITE3_WRITE_CONNECTION": "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix-dotfile",
    "NANOS_SQLITE3_READ_CONNECTION": "file:db/nanos.sqlite3?mode=ro&_pragma=busy_timeout=5000&vfs=unix-dotfile",
    "NANOS_SQLITE3_MAX_GOROUTINES_WRITE": "1",
    "NANOS_SQLITE3_MAX_GOROUTINES_READ": "1",
    "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE": "100",
    "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ": "100",
    "NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS": "0",
    "NANOS_SQLITE3_READ_ALL_DATA": "",
    "NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS": "",
    "GODEBUG": "gctrace=0,scavtrace=0",
    "GOMAXPROCS": ""
  },
  "Debugflags": [
    "reboot_on_exit"
  ]
}
```
</details>

```
...
2023/08/06 16:27:31 R/W goroutines work - STARTING

frame trace: 
ffffc00000cbfec0:   ffffffff80090519	(filesystem_put_node + 0000000000000019/000000000000005b)
ffffc00000cbfee0:   ffffffff800adcab	(utime_internal + 00000000000000eb/000000000000016a)
ffffc00000cbff40:   ffffffff800aee37	(utimes + 0000000000000187/0000000000000424)
ffffc00000cbffb0:   ffffffff800e5f98	(syscall_handler + 00000000000002e8/000000000000061b)

loaded klibs: 
assertion ctx == m->turn failed at /home/circleci/project/src/kernel/mutex.c:263 (IP 0xffffffff800491af)  in mutex_unlock(); halt
```

The same service setup under **linux** may report (expected and not important, just that it doesn't crash/fail)

<details><summary><b>linux</b> run</summary>

```sh
$ mkdir db
$ ./nanos-sqlite.sh
```

- `nanos-sqlite.sh` content

```sh
#!/usr/bin/env bash

set -o errexit
set -o pipefail

# export NANOS_SQLITE3_VFS=unix # default
#
export NANOS_SQLITE3_VFS=unix-dotfile
# export NANOS_SQLITE3_VFS=unix-excl
#
export NANOS_SQLITE3_WRITE_CONNECTION="file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=${NANOS_SQLITE3_VFS:-unix}"
export NANOS_SQLITE3_READ_CONNECTION="file:db/nanos.sqlite3?mode=ro&_pragma=busy_timeout=5000&vfs=${NANOS_SQLITE3_VFS:-unix}"
export NANOS_SQLITE3_MAX_GOROUTINES_WRITE=1
export NANOS_SQLITE3_MAX_GOROUTINES_READ=1
export NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE=100
export NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ=100
export NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS=0
export NANOS_SQLITE3_READ_ALL_DATA=
export NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS=
export GOMAXPROCS=1

./nanos-sqlite
```

</details>

```
...
2023/08/06 16:26:59 R/W goroutines work - STARTING
2023/08/06 16:26:59 WRITE: GW(goroutine_0) - stmt.Exec(goroutine_0-1) err("database is locked (5) (SQLITE_BUSY)")
2023/08/06 16:26:59 WRITE: GW(goroutine_0) - stmt.Exec(goroutine_0-2) err("database is locked (5) (SQLITE_BUSY)")
.....
2023/08/06 16:27:05 R/W goroutines work - COMPLETED
...
```
---
Now back to (522 error).  First lets  chang vfs back to its default value `vfs=unix` (or delete it from config)

While there are different ways of reproducing this, the fastest is to set the number of `write goroutines >` 1 or `increasing --smp` or `GOMAXPROCS`. The following config should be enough to trigger the error. All equivalent configs when run on linux work correctly.

<details><summary><b>ops</b> config</summary>

```sh
$ ops volume delete nanos-sqlite && ops volume create nanos-sqlite -s 10G
$ ops run -c config_522.json nanos-sqlite --smp=1 -f -n
```

- `config_522.json`

```json
{
  "BaseVolumeSz": "32m",
  "Args": [],
  "Mounts": {
    "nanos-sqlite": "db"
  },
  "Dirs": [],
  "RunConfig": {
    "Memory": "8G"
  },
  "Env": {
    "NANOS_SQLITE3_WRITE_CONNECTION": "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix",
    "NANOS_SQLITE3_READ_CONNECTION": "file:db/nanos.sqlite3?mode=ro&_pragma=busy_timeout=5000&vfs=unix",
    "NANOS_SQLITE3_MAX_GOROUTINES_WRITE": "2",
    "NANOS_SQLITE3_MAX_GOROUTINES_READ": "1",
    "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_WRITE": "100",
    "NANOS_SQLITE3_PER_GOROUTINE_MAX_ITEMS_READ": "100",
    "NANOS_SQLITE3_RW_RANDOM_WAIT_MILLIS": "0",
    "NANOS_SQLITE3_READ_ALL_DATA": "",
    "NANOS_SQLITE3_USE_MULTIPLE_CONNECTIONS": "",
    "GODEBUG": "gctrace=0,scavtrace=0",
    "GOMAXPROCS": ""
  },
  "Debugflags": [
    "reboot_on_exit"
  ]
}
```
</details>


<details><summary><b>disk I/O error (522)</b> output example</summary>

```sh
running local instance
booting /home/rinor/.ops/images/nanos-sqlite ...
en1: assigned 10.0.2.15
en1: assigned FE80::88EC:77FF:FED8:AF64
2023/08/06 18:30:31 
2023/08/06 18:30:31 Connection Write String: "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix"
2023/08/06 18:30:31 Connection Read String : "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix"
2023/08/06 18:30:31 
2023/08/06 18:30:31 Use mutiple connections: false
2023/08/06 18:30:31 
2023/08/06 18:30:31 Write goroutines max: 2
2023/08/06 18:30:31 Read  goroutines max: 1
2023/08/06 18:30:31 
2023/08/06 18:30:31 Write items per goroutine: 100
2023/08/06 18:30:31 Read  items per goroutine: 100
2023/08/06 18:30:31 
2023/08/06 18:30:31 R/W Random Wait Millis : [0,0)
2023/08/06 18:30:31 ReadAllData at the end : false
2023/08/06 18:30:31 
2023/08/06 18:30:31 SQLITE_VERSION()  : 3.41.2
2023/08/06 18:30:31 SQLITE_SOURCE_ID(): 2023-03-22 11:56:21 0d1fc92f94cb6b76bffe3ec34d69cffde2924203304e8ffc4155597af0c191da
2023/08/06 18:30:31 
2023/08/06 18:30:31 R/W goroutines work - STARTING
2023/08/06 18:30:31 READ: GR(0) - row.Scan(79) err("disk I/O error (522)")
2023/08/06 18:30:31 READ: GR(0) - row.Scan(78) err("disk I/O error (522)")
2023/08/06 18:30:31 READ: GR(0) - row.Scan(77) err("disk I/O error (522)")
2023/08/06 18:30:31 READ: GR(0) - row.Scan(76) err("disk I/O error (522)")
.....
2023/08/06 18:30:32 READ: GR(0) - row.Scan(3) err("disk I/O error (522)")
2023/08/06 18:30:32 READ: GR(0) - row.Scan(2) err("disk I/O error (522)")
2023/08/06 18:30:32 READ: GR(0) - row.Scan(1) err("disk I/O error (522)")
2023/08/06 18:30:32 R/W goroutines work - COMPLETED
2023/08/06 18:30:32 
2023/08/06 18:30:32 DEFER - SETUP - stmt_main_select(0xc000108120) err(%!q(<nil>))- closed
2023/08/06 18:30:32 DEFER - SETUP - stmt_main_insert(0xc000108090) err(%!q(<nil>)) - closed
2023/08/06 18:30:32 DEFER - SETUP - db(0xc000019380) err(%!q(<nil>)) - closed
2023/08/06 18:30:32 
2023/08/06 18:30:32 DEFER - MAIN - ...will terminate in 5s
2023/08/06 18:30:37 DEFER - MAIN - DONE...
```
</details>

**Comment by francescolavra** [comment]
The kernel assertion failure when using `vfs=unix-dotfile` is fixed by #1920.
The SQLITE_IOERR_SHORT_READ error when using `vfs=unix` is solved by #1921.

**Comment by rinor** [comment]
tested both prs merged/rebased with main and:

---

- `vfs=unix` - is much more stable, yet to perform more tests, but so far it looks ok

---

- `vfs=unix-dotfile` - is a bit better, but still the behavior is strange/unstable.

```
booting /home/rinor/.ops/images/nanos-sqlite ...
en1: assigned 10.0.2.15
en1: assigned FE80::5C39:AAFF:FE7E:C98F
2023/08/17 15:24:28 
2023/08/17 15:24:28 Connection Write String: "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix-dotfile"
2023/08/17 15:24:28 Connection Read String : "file:db/nanos.sqlite3?mode=rwc&_pragma=journal_mode=WAL&_pragma=synchronous=NORMAL&_pragma=busy_timeout=5000&_pragma=locking_mode=NORMAL&_pragma=auto_vacuum=full&vfs=unix-dotfile"
2023/08/17 15:24:28 
2023/08/17 15:24:28 Use mutiple connections: false
2023/08/17 15:24:28 
2023/08/17 15:24:28 Write goroutines max: 2
2023/08/17 15:24:28 Read  goroutines max: 1
2023/08/17 15:24:28 
2023/08/17 15:24:28 Write items per goroutine: 100
2023/08/17 15:24:28 Read  items per goroutine: 100
2023/08/17 15:24:28 
2023/08/17 15:24:28 R/W Random Wait Millis : [0,0)
2023/08/17 15:24:28 ReadAllData at the end : false
2023/08/17 15:24:28 
2023/08/17 15:24:28 SQLITE_VERSION()  : 3.41.2
2023/08/17 15:24:28 SQLITE_SOURCE_ID(): 2023-03-22 11:56:21 0d1fc92f94cb6b76bffe3ec34d69cffde2924203304e8ffc4155597af0c191da
2023/08/17 15:24:28 
2023/08/17 15:24:28 R/W goroutines work - STARTING
2023/08/17 15:24:28 WRITE: GW(goroutine_1) - stmt.Exec(goroutine_1-1) err("disk I/O error (5898)")
2023/08/17 15:24:28 WRITE: GW(goroutine_1) - stmt.Exec(goroutine_1-2) err("disk I/O error (5898)")
2023/08/17 15:24:28 WRITE: GW(goroutine_1) - stmt.Exec(goroutine_1-3) err("disk I/O error (5898)")
2023/08/17 15:24:28 WRITE: GW(goroutine_1) - stmt.Exec(goroutine_1-4) err("disk I/O error (5898)")
2023/08/17 15:24:28 WRITE: GW(goroutine_1) - stmt.Exec(goroutine_1-5) err("disk I/O error (5898)")
2023/08/17 15:24:29 WRITE: GW(goroutine_0) - stmt.Exec(goroutine_0-5) err("disk I/O error (5898)")
-
- -> most of the times it gets stuck like this (after 1 or more write attempts)
-
```

**Comment by francescolavra** [comment]
I was testing with NANOS_SQLITE3_MAX_GOROUTINES_WRITE=1 and I couldn't reproduce the SQLITE_IOERR_DELETE_NOENT error. Now with NANOS_SQLITE3_MAX_GOROUTINES_WRITE=2 I can see the error; it will be fixed by #1924.

**Comment by rinor** [comment]
tested and works as expected. Thank you.
